### DISCUSSION TITLE
Using Benthos as a Conduit Processor

### DISCUSSION THREAD

**Discussion by nickchomey** [ConduitIO/conduit#1614] [Ideas]
I'm relatively new to these sorts of tools, but it seems to me that Conduit and Benthos are somewhat redundant as they are both stream processors. As such, it seems somewhat silly to use them together (as with the PoC Benthos Connector - https://github.com/conduitio-labs/conduit-connector-benthos/issues/4) - better to choose one.

The main difference between them seems to be that Conduit is much more focused on CDC from data sources/stores/bases while Benthos is much more focused on the actual pipeline processing/transformation - it has many [dozens of processors](https://www.benthos.dev/docs/components/processors/about) while Conduit has only a handful. Their processors also allow for enrichment via Sql queries, nats kv etc...

It seems to me that the universal/OpenCDC of Conduit is far more fundamental/important, since a pipeline ultimately needs to start from some data source, and should therefore be used as the main tool. But it would be a shame to not leverage the immense processing power of Benthos.

So, what I'm thinking is that rather than use Benthos as a Conduit Source/Destination, as was attempted in this repo, why not just embed it's pipeline processors into Conduit as a [Standalone Processor](https://conduit.io/docs/processors/standalone/)? It could have some sort of Benthos bloblang mechanism for choosing the desired Benthos processors.

This would allow Conduit to focus on its strength of CDC, while leveraging Benthos' strength in stream processing. You could, of course, always make other custom processors in Go or JavaScript to suit your needs (or probably even use existing Benthos custom processors).

It's a topic that has been brought up various times in Benthos' Github and Discord, and they're generally responded to with the following links:

Apparently this can be used to embed Benthos into a golang app/binary
https://pkg.go.dev/github.com/benthosdev/benthos/v4/public/service#example-package-StreamBuilderConfig

One more example of that api here https://github.com/benthosdev/benthos/issues/1727#issuecomment-1423210719

And here's a repo that apparently has relevant examples
https://github.com/benthosdev/benthos-plugin-example

Thoughts?

**Comment by nickchomey** [comment]
It was briefly mentioned in the original issue in that this would face challenges as standalone processors need to be wasm.

Here is an article from the Benthos dev about how to build Benthos in wasm. Perhaps this can be done here?

https://www.benthos.dev/blog/2019/05/27/compiling-benthos-to-wasm/

**Comment by lovromazgon** [comment]
This is definitely an interesting idea. While Conduit provides a set of [builtin](https://conduit.io/docs/processors/builtin/) processors including a [JavaScript](https://conduit.io/docs/processors/builtin/custom.javascript) processor and a [Go SDK](https://conduit.io/docs/processors/standalone/building) for standalone processors, it's currently missing an easy way to enrich data in the pipeline. This is the part where I think Benthos could provide a lot of value. It could also be valuable to users who are already familiar with Benthos and would like to use Bloblang to process their data.

Sadly, we'll hit an issue if we try to tackle this as a standalone processor. The thing is that standalone processors currently run in a constrained [WASM sandbox](https://conduit.io/docs/processors/standalone/how-it-works#wasi-support-and-limitations) and don't have access to the file system or network. This makes enriching pretty much impossible, with or without Benthos. To solve this we'd have to bind system calls for the standard `net` package, say using something like [`github.com/dispatchrun/net`](https://github.com/dispatchrun/net), although last time I played around with that package it didn't seem stable enough. I found issues with reading the response body and I got stuck with configuring HTTPS, as the WASM plugin would need to have access to the certificate store. Additionally, the lib only provides bindings for UNIX systems, which is ok for a start, but it would break the Windows release of Conduit. Doing this right is not trivial, but in the long run, we definitely want to add this functionality, as it would be a very powerful addition.

In other words, implementing a Benthos _standalone_ processor would require significant effort to enable the needed functionality. However, there's another way - a _builtin_ processor. A builtin processor is running inside of Conduit directly (regular Go code), so it doesn't have any of the WASM sandbox limitations. This approach could work, although we don't plan to tackle this in the next few months, so I'll give you some pointers, in case you want to give it a go yourself. If you wanted to create a fully functional Benthos processor, you would have to create your own entrypoint (like we have [here](https://github.com/ConduitIO/conduit/blob/main/cmd/conduit/main.go)) and add the Benthos processor to the global map of [builtin processors](https://github.com/ConduitIO/conduit/blob/6d59f708ad221b9b462c1c2d2278e690e8011c64/pkg/plugin/processor/builtin/registry.go#L36) before calling [`conduit.Serve`](https://github.com/ConduitIO/conduit/blob/6d59f708ad221b9b462c1c2d2278e690e8011c64/pkg/conduit/entrypoint.go#L31). The Benthos processor configuration would have to be provided by the user as a string, which will make it a bit awkward to write it in the [pipeline configuration file](https://conduit.io/docs/pipeline-configuration-files/getting-started), as it will basically be a yaml formatted string inside a yaml file. This doesn't mean it can't be done, to prove the value of such a processor.

Let us know if this was helpful and if we can provide any further guidance, in case you decide to give it a go.

**Reply by nickchomey** [reply]
Thanks very much for the insights!

I mentioned in a separate reply in this discussion that Benthos can be compiled to WASM. Might that allow for making a Benthos standalone processor? The processors that depend on filesystem or network access wouldn't work, but there's still dozens of other pure processors that surely would still be useful. 

Though, either way, it seems like a built-in benthos processor with pure golang is a better option as it would avoid the added complexity of building it in wasm, as well as the loss of functionality. 

I'll try to take a look at the code you shared in the coming week or so and see if I can get something working. I'll write back for guidance if I get stuck. 

For the benthos yaml, it seems like it might be possible to do something like the [JavaScript built in processor](https://conduit.io/docs/processors/builtin/custom.javascript) - either have a string field or, better yet, a file path to the yaml file. In fact, I could probably just use the JS processor as a template to get started. Do you think that could work? 

Thanks again for your prompt attention!

**Reply by lovromazgon** [reply]
It seems that compiling Benthos to WASM shouldn't be an issue, I just wonder what will happen if Benthos tries to expose some HTTP endpoints and if it experiences an error. Maybe it's not an issue, or maybe Benthos can be configured to disable the API, I'm not sure, it would have to be tried out. If you do get it running, the "pure" Benthos processors, which don't reach out to the file system or network, should work perfectly fine in a standalone processor.

Using the JavaScript processor as a template sounds reasonable. Also, the approach to get the Benthos YAML sounds reasonable, essentially following the approach in the JavaScript processor using `script` and `script.path` 👍

**Reply by nickchomey** [reply]
I will try the built-in processor first, as it should be simpler and would be full-featured. I hope to get to it in the next week or so.

**Reply by nickchomey** [reply]
I was just reading more about WASM and WASI and thought I'd share some more info for others' sake. [WASI](https://github.com/WebAssembly/WASI) is an extension of WASM (which was mostly focused on being used in browsers), which aims to add support for network, filesystem and many more APIs. 

The [conduit docs](https://conduit.io/docs/processors/standalone/how-it-works/#wasi-support-and-limitations) state that the WASI implementation in Standalone Plugins uses WASI Preview 1. This is fairly limited. But [Preview 2](https://github.com/WebAssembly/WASI/tree/main/preview2) is in development right now and will have support for HTTP, filesystem and more. 

So, it stands to reason that Benthos could be implemented as a standalone plugin when Conduit adopts WASI preview 2. 

I'll still probably look into making a built-in processor out of curiosity, but might not follow through as it isn't a huge priority for me at the moment - more of a nice-to-have. 

Here's all of the [WASI API proposals](https://github.com/WebAssembly/WASI/blob/main/Proposals.md) and their current status. It looks like it will be quite comprehensive someday!

**Reply by nickchomey** [reply]
Benthos has been acquired by Redpanda. https://techcrunch.com/2024/05/30/redpanda-acquires-benthos-to-expand-its-end-to-end-streaming-data-platform

It is now known as Redpanda Connect. https://github.com/redpanda-data/connect

Its not yet clear what their plans are for it, but there's a [livestream](https://www.youtube.com/live/X8nVdUuWZ80) in a few hours about it. I'll update this comment with relevant info afterwards

**Reply by hariso** [reply]
Yup, we heard the news too, and will be following how things evolve.:) It looks like the license remained the same.

**Reply by nickchomey** [reply]
Sounds like there aren't any meaningful changes so far. The plugins have been broken out from core, which would resemble Conduit's architecture where you pick and choose what you want to use (as opposed to having a single massive binary with all plugins). Their license changed from MIT to Apache, but that's not a big deal. 

2 plugins changed to a commercial license though - Splunk and Snowflake. I'm not sure that matters too much as they're just destination connectors, not processors - and Conduit has Snowflake already. 

[WarpStream forked Benthos to Bento](https://www.warpstream.com/blog/announcing-bento-the-open-source-fork-of-the-project-formerly-known-as-benthos), but I suspect it is somewhat premature. Better to wait until there's a genuine rug pull to cause a rift in the "community"

It might be a good opportunity for Conduit/Meroxa to get some attention if a Benthos/Redpanda Connect built-in processor could be made, and perhaps move it to a standalone processor later when WASI p2 is ready?

Edit: this comment from the Benthos/redpanda [press release](https://www.businesswire.com/news/home/20240530590089/en/Redpanda-Acquires-Benthos-to-Deliver-a-Complete-End-to-End-Streaming-Data-Platform) suggests something to this effect is being planned...?

> Meroxa is excited to partner with Redpanda to bring multi-language SDK support for the Redpanda Connect ecosystem via our popular open source project Conduit," said DeVaris Brown, founder and CEO of Meroxa. "As an existing Redpanda customer, we are excited to deepen our partnership to bring to the world the best developer experience for the connector ecosystem with Meroxa and Redpanda Connect."

**Reply by hariso** [reply]
@nickchomey Thanks for the summary! Meroxa is keeping an eye on all of this. I hope that soon we'll have something that is great for the companies and the open-source community.:)

I read the news about WarpStream. I believe they they didn't want to be caught by surprise by a sudden license change (which could affect them a lot, since it looks like they use it heavily).

> It might be a good opportunity for Conduit/Meroxa to get some attention if a Benthos/Redpanda Connect built-in processor could be made, and perhaps move it to a standalone processor later when WASI p2 is ready?

That's definitely an option, a good one too. As Lovro said, we didn't plan to get onto that ourselves in the near future, but we can re-evaluate .:)

**Reply by mihaitodor** [reply]
Hey folks, this looks like a cool idea! Since wasm was mentioned, I thought I'd also suggest looking into having Bloblang as a WASM processor as a quick feature which might be fun to play with. Here's an example of how that can be built as WASM: https://github.com/mihaitodor/benthos-playground/tree/main/wasm/blobl-editor. Unfortunately, there's an issue currently in here https://github.com/redpanda-data/benthos/issues/20 which breaks the build, so that will have to get fixed, but `v4.11.0` should work for experiments. I haven't checked which version introduced the build failure.

**Reply by nickchomey** [reply]
Hi @mihaitodor, I recognize you as one of the primary contributors to Benthos! Thanks for your work there and  interest in this! 

When you say a bloblang wasm processor for Conduit, would that be something different than what was already proposed here (a Benthos processor for Conduit - which uses wasm for its standalone (non-built-in) processors? 

Because, as was detailed above, it's likely that wasi preview 1 - used by conduit for the wasm processors - is not sufficient to make use of all of benthos' functionality. So, in the meantime it was proposed to make a built-in golang processor, which would definitely work. Or do you think it actually could work? 

Or do you mean something different from Benthos when you say "bloblang"? Would it be just a subset of benthos? What would it do in the context of a Conduit pipeline?

**Reply by mihaitodor** [reply]
@nickchomey Happy to help the community and curious to see how Conduit will evolve!

> When you say a bloblang wasm processor for Conduit, would that be something different than what was already proposed here (a Benthos processor for Conduit - which uses wasm for its standalone (non-built-in) processors?

Benthos has an embedded DSL called Bloblang (see the docs [here](https://docs.redpanda.com/redpanda-connect/guides/bloblang/about/)) which can be used in the `mapping` / `mutation` processors as well as in [interpolated / mapping fields](https://docs.redpanda.com/redpanda-connect/configuration/interpolation/#bloblang-queries). It was created as a Benthos-native language which can be used for succinct data queries and transformations which may not be as easy to implement and maintain as Javascript in the `javascript` processor for example. Also, it should be more efficient given that it's parsed and transpiled directly to the underlying Go functions on startup. That's the sales pitch anyway :) I'd be happy to hear other people try it out and see if they like it. I have some non-trivial examples that I maintain [here](https://github.com/mihaitodor/benthos-playground/tree/main/bloblang), but the documentation is a good place to start.

> So, in the meantime it was proposed to make a built-in golang processor, which would definitely work. Or do you think it actually could work?

Yep, that should work too and it offers you a means of embedding a Benthos _processors pipeline_ as a processor in your Conduit pipeline. Have a look at [this `StreamBuilder` example](https://pkg.go.dev/github.com/redpanda-data/benthos/v4@v4.28.1/public/service#example-package-StreamBuilderPush). You can use the `AddProducerFunc()` and `AddConsumerFunc` APIs (or `AddBatchProducerFunc()` and `AddBatchConsumerFunc()` if you want to process message batches) to push data to the Benthos pipeline and read the results. This is way more generic, since the `mapping` / `mutation` processors are just a subset of all the [other processors](https://docs.redpanda.com/redpanda-connect/components/processors/about/) which are available to use out of the box and there are APIs to write your own custom ones (see the example [here](https://pkg.go.dev/github.com/redpanda-data/benthos/v4@v4.28.1/public/service#example-package-ProcessorPlugin)). The only caveat is that you'll likely have to import them selectively when compiling to WASM, since I imagine not all of them support WASM (depending on the underlying 3rd party libraries). You can see [here](https://github.com/redpanda-data/connect/blob/main/public/bundle/free/package.go) a list of individual imports that you can pull in to bring in processors (those also register inputs and outputs but I think you won't need to use those initially) selectively.

**Reply by nickchomey** [reply]
Yes, I think we're aware of what bloblang is and does. But what I'm trying to understand is what a bloblang-only (without the rest of benthos) standalone processor for Conduit would offer/allow?

And, yes, a Benthos processors pipeline as a Conduit processor is exactly what was proposed and already discussed in considerable detail here. I'm glad you've been able to confirm what was previously suspected - that Benthos' functionality would be limited in wasm/wasi until wasi evolves further (likely preview 2, linked to above, will support the full range of benthos capabilities in wasi).

So, we're left with the 2 options that have already been discussed - wait for wasi preview 2, or build a Benthos pipeline processor as a Conduit built-in processor. It would be great if you or others from the Benthos/ Redpanda team could assist with this!

**Reply by mihaitodor** [reply]
> Yes, I think we're aware of what bloblang is and does. But what I'm trying to understand is what a bloblang-only (without the rest of benthos) standalone processor for Conduit would offer/allow?

I noticed you have a [`javascript` processor](https://conduit.io/docs/processors/builtin/custom.javascript) and, in many cases, people don't need to write very complex transformations which require multiple processors. Basically, it would be a shortcut for someone who just wants to write a Bloblang transformation and use it with either Benthos or Conduit. However, they can achieve the same with a Benthos processor which contains just a `mapping` / `mutation` processor in it. My initial thought was that it would be easier to include Bloblang as WASM if you want to go down that route rather than a Benthos stream. But, if you're willing to import Benthos as a library in your Go code, then it makes sense to go for the stream version and support multiple processors.

> It would be great if you or others from the Benthos/ Redpanda team could assist with this!

Feel free to raise a matching enhancement proposal on https://github.com/redpanda-data/connect so we can track it. I'm a bit caught up right now, but happy to keep in touch about this.

**Reply by nickchomey** [reply]
@mihaitodor yes, JavaScript is another option or processing in conduit. I agree that bloblang would be easier, though can it do much without having all of the other benthos processors available? 

I think having the full range of benthos processors (but not necessarily the inputs and outputs) within a conduit pipeline would be ideal for most people, as it would combine the powerful CDC capabilities of Conduit with the processing flexibility and simplicity of benthos and bloblang.

I'll create an issue at the benthos repo in a bit. Hopefully this can happen at some point!

**Reply by mihaitodor** [reply]
> though can it do much without having all of the other benthos processors available?

You'd get the full list of currently available [functions](https://docs.redpanda.com/redpanda-connect/guides/bloblang/functions/) and [methods](https://docs.redpanda.com/redpanda-connect/guides/bloblang/methods/), which I think is on par with what you can do with Javascript, although there probably are more Javascript libraries and snippets out there that you can just import and use. Would be cool to know what else people would like to have available in bloblang.

> I'll create an issue at the benthos repo in a bit. Hopefully this can happen at some point!

Thank you! ❤️

**Reply by nickchomey** [reply]
Ah ok, very cool. Perhaps I'll try to make a bloblang wasm processor sometime soon, in lieu of a full processor pipeline plugin.

**Reply by nickchomey** [reply]
As mentioned in #1646, Wazero has no plans whatsoever to add support for WASI preview 2 (or any other preview until 1.0 and wasi is in w3c spec. So, implementing Benthos as a standalone processor seems to be impossible. Perhaps at some point i'll look at doing so as a built-in processor, as discussed above
